File: src/main/java/com/github/hakobera/jskinect/KinectTracker.java
package com.github.hakobera.jskinect;

import org.openkinect.processing.Kinect;

import processing.core.PApplet;
import processing.core.PConstants;
import processing.core.PImage;
import processing.core.PVector;

public class KinectTracker {
	
	private static final int KW = 640;
	private static final int KH = 480;
	
	private PApplet processing;
	private Kinect kinect;
	
	private int threshold = 600;

	/**
	 * トラッカーの位置
	 */
	private PVector loc;

	/**
	 * 位置補完後のトラッカーの位置
	 */
	private PVector lerpedLoc;
	
	private int[] depth;
	
	private PImage display;
	 
	/**
	 * コンストラクタ
	 * 
	 * @param processing processing 
	 * @param kinect Kinect 操作オブジェクト
	 */
	public KinectTracker(PApplet processing, Kinect kinect) {
		this.processing = processing;
	    this.kinect = kinect;
	    
		kinect.start();
	    kinect.enableRGB(false);
	    kinect.enableDepth(true);
	    kinect.processDepthImage(true);

	    display = processing.createImage(KW, KH, PConstants.RGB);

	    loc = new PVector(0, 0);
	    lerpedLoc = new PVector(0, 0);
	}
	 
	public void track() {
		depth = kinect.getRawDepth();

	    if (depth == null) {
	      return;
	    }

	    float sumX = 0;
	    float sumY = 0;
	    float count = 0;

	    for (int x = 0; x < KW; ++x) {
	    	for (int y = 0; y < KH; y++) {
	    		int offset = KW - x - 1 + y * KW;
	    		int rawDepth = depth[offset];

	    		if (rawDepth < threshold) {
	    			sumX += x;
	    			sumY += y;
	    			count++;
	    		}
	    	}
	    }

	    if (count != 0) {
	    	loc = new PVector(sumX/count, sumY/count);
	    }

	    lerpedLoc.x = PApplet.lerp(lerpedLoc.x, loc.x, 0.3f);
	    lerpedLoc.y = PApplet.lerp(lerpedLoc.y, loc.y, 0.3f);
	}

	public PVector getLerpedPos() {
		return lerpedLoc;
	}

	public PVector getPos() {
		return loc;
	}
	  
	public void display() {
		PImage img = kinect.getDepthImage();
	    if (depth == null || img == null) {
	    	return;
	    }
	    
	    display.loadPixels();
	    for (int x = 0; x < KW; ++x) {
	    	for (int y = 0; y < KH; ++y) {
	    		int offset = KW - x - 1 + y * KW;
	    		int rawDepth = depth[offset];
	        
	    		int pix = x + y * display.width;
	    		if (rawDepth < threshold) {
	    			display.pixels[pix] = processing.color(150, 50, 50);
	    		} else {
	    			display.pixels[pix] = img.pixels[offset];
	    		}
	    	}
	    }
	    display.updatePixels();
	    
	    processing.image(display, 0, 0);
	}
	  
	public void quit() {
		kinect.quit();
	}
	  
	public void setThreshold(int t) {
		threshold = t;
	}
	  
	public int getThreshold() {
		return threshold;
	}
	  
}

File: js/lib/kinectTracker.js
/**
 * 深度しきい値最小
 *
 * @property
 * @type {number}
 */
var _thresholdMin = 400;

/**
 * 深度しきい値最大
 *
 * @property
 * @type {number}
 */
var _thresholdMax = 500;

/**
 * 追跡座標
 *
 * @property
 * @type {PVector}
 */
var _location;

/**
 * 補完後追跡座標
 *
 * @property
 * @type {PVector}
 */
var _lerpedLocation;

/**
 * 表示用イメージ
 *
 * @private
 * @type {PImage}
 */
var _display;

/**
 * @private
 * @type {PApplet}
 */
var _processing;

/**
 * @private
 * @type {Kinect}
 */
var _kinect;

/**
 * 取得できる画像の横解像度(px)
 *
 * @private
 * @type {number}
 */
var _width;

/**
 * 取得できる画像の縦解像度(px)
 *
 * @private
 * @type {number}
 */
var _height;

/**
 * 深度情報配列
 *
 * @private
 * @type {Array.<number>}
 */
var _depth;

/**
 * 引数がしきい値内に収まっているかどうかを判定して返します。
 *
 * @param {number} v 判定する値
 * @return {boolean} しきい値内にある場合、true
 * @private
 */
function isInRange(v) {
    return _thresholdMin < v && v < _thresholdMax;
}

/*
 * import
 */
var PVector = Packages.processing.core.PVector;

/**
 * @module KinectTracker
 */
exports = {

    /**
     * PApplet， Kinect オブジェクトと関連付け。
     *
     * @param {PApplet} processing procesing アプレット
     * @param {Kinect} kinect Kinect 接続オブジェクト
     */
    bind: function(processing, kinect) {
        _processing = processing;
        _kinect = kinect;
        _width = kinect.width() | 0;
        _height = kinect.height() | 0;
        _display = _processing.createImage(_width, _height, Packages.processing.core.PConstants.RGB);
        _location = new PVector(0, 0);
        _lerpedLocation = new PVector(0, 0);
    },

    /**
     * しきい値内の手の位置をトラッキングします。
     */
    track: function(render) {
        var img = _kinect.getDepthImage(),
            sumX = 0,
            sumY = 0,
            count = 0,
            detectColor = _processing.color(150, 50, 50),
            x, y, yOffset, offset, rawDepth, pt, src, dest, c;

        _depth = _kinect.getRawDepth();

        if (render) {
	        _display.loadPixels();
	        src = img.pixels;
	        dest = _display.pixels;
        }
	    for (y = 0; y < _height; y+=1) {
            yOffset = y * _width;
            for (x = 0; x < _width; x+=1) {
                offset = (_width - x - 1) + yOffset | 0; // 左右反転を元に戻す
                rawDepth = _depth[offset];
                pt = x + yOffset;
                if (isInRange(rawDepth)) {
                    sumX += x;
                    sumY += y;
                    count += 1;
                }
                
                if (render) {
                    if (isInRange(rawDepth)) {
                        c = detectColor;
                    } else {
                        c = src[offset];
                    }
                	dest[pt] = c;
                }
            }
        }
	    if (render) {
	        _display.updatePixels();
	        _processing.image(_display, 0, 0);
	    }
	    
        if (count !== 0) {
            _location.x = sumX / count;
            _location.y = sumY / count;
        }

        _lerpedLocation.x = _processing.lerp(_lerpedLocation.x, _location.x, 0.3);
        _lerpedLocation.y = _processing.lerp(_lerpedLocation.y, _location.y, 0.3);
    },

    /**
     * 補完後のトラッキング座標を返します。
     *
     * @return {PVector} 補完後のトラッキング座標
     */
    getPos: function() {
        return _lerpedLocation;
    },

    /**
     * しきい値を設定します。
     * 
     * @param {number} t しきい値
     */
    setThreshold: function(t) {
        _thresholdMin = t;
    },

    /**
     * 設定されているしきい値を返します。
     *
     * @return {number} 設定されてるしきい値
     */
    getThreshold: function() {
        return _thresholdMin;
    }
    
};
